src/util/http/websocket/QueryToSocketDistributor.cpp

```diff
@@ -4,52 +4,33 @@
 
 #include "util/http/websocket/QueryToSocketDistributor.h"
 
-#include <boost/asio/associated_cancellation_slot.hpp>
+#include <boost/asio/as_tuple.hpp>
 #include <boost/asio/use_awaitable.hpp>
 
 #include "util/Log.h"
 
 namespace ad_utility::websocket {
 
-template <typename CompletionToken>
-auto QueryToSocketDistributor::waitForUpdate(CompletionToken&& token) {
-  auto id = counter++;
-  auto initiate = [strand = strand_, wakeupCalls = wakeupCalls_,
-                   id]<typename Handler>(Handler&& self) mutable {
-    auto sharedSelf = std::make_shared<Handler>(std::forward<Handler>(self));
-    auto cancellationSlot = net::get_associated_cancellation_slot(
-        *sharedSelf, net::cancellation_slot());
-
-    net::dispatch(strand, [wakeupCalls, sharedSelf, id]() mutable {
-      wakeupCalls->emplace_back(IdentifiableFunction{
-          .function_ = [sharedSelf =
-                            std::move(sharedSelf)]() { (*sharedSelf)(); },
-          .id_ = id});
-    });
-
-    if (cancellationSlot.is_connected()) {
-      cancellationSlot.assign([strand = std::move(strand),
-                               wakeupCalls = std::move(wakeupCalls),
-                               id](net::cancellation_type) mutable {
-        // Make sure wakeupCalls_ is accessed safely.
-        net::dispatch(strand, [wakeupCalls = std::move(wakeupCalls), id]() {
-          std::erase_if(*wakeupCalls,
-                        [id](auto idFunc) { return idFunc.id_ == id; });
-        });
-      });
+net::awaitable<void> QueryToSocketDistributor::waitForUpdate() {
+  try {
+    co_await infiniteTimer_.async_wait(net::use_awaitable);
+    AD_THROW("Infinite timer expired. This should not happen");
+  } catch (boost::system::system_error& error) {
+    if (error.code() != net::error::operation_aborted) {
+      throw;
     }
```

Comment:
It's possible to avoid the try catch block in this case, but when using coroutines the errors are propagated via exceptions, so rethrowing is pretty simple instead of having to rebuild a system error using the acquired error code. I'd suggest we discuss this a small change for a future PR, because there are several ways to do it and this PR is already big enough